src/razor/src/completion/completionHandler.ts

```diff
@@ -217,54 +259,39 @@ export class CompletionHandler {
             return csharpCompletions;
         } catch (error) {
             this.logger.logWarning(`${CompletionHandler.completionEndpoint} failed with ${error}`);
+        } finally {
+            if (provisionalTextEdit && virtualDocument.removeProvisionalDot()) {
+                const removeDot = true;
+                await this.ensureProvisionalDotUpdatedInCSharpDocument(virtualDocument.uri, absoluteIndex, removeDot);
+            }
         }
 
         return CompletionHandler.emptyCompletionList;
     }
 
-    // Provides 'provisional' C# completions.
-    // This happens when a user types '.' after an object. In that case '.' is initially in
-    // html document and not generated C# document. To get correct completions as soon as the user
-    // types '.' we need to
-    // 1. Temporarily add '.' to projected C# document at the correct position (projected position)
-    // 2. Make sure projected document is updated on the Roslyn server so Roslyn provides correct completions
-    // 3. Invoke Roslyn/C# completion and return that to the Razor LSP server.
-    //    NOTE: currently there is an issue (see comments in code below) causing us to invoke vscode command
-    //    rather then the Roslyn command
-    // 4. Remove temporarily (provisionally) added '.' from the projected C# buffer.
-    // 5. Make sure the projected C# document is updated since the user will likely continue interacting with this document.
-    private async provideCSharpProvisionalCompletions(
-        triggerCharacter: string | undefined,
-        virtualDocument: CSharpProjectedDocument,
-        projectedPosition: Position
-    ) {
-        const absoluteIndex = CompletionHandler.getIndexOfPosition(virtualDocument, projectedPosition);
-        if (absoluteIndex === -1) {
-            return CompletionHandler.emptyCompletionList;
-        }
-
-        try {
-            // temporarily add '.' to projected C# document to ensure correct completions are provided
-            virtualDocument.addProvisionalDotAt(absoluteIndex);
-            await this.ensureProjectedCSharpDocumentUpdated(virtualDocument.uri);
-
-            // Current code has to execute vscode command vscode.executeCompletionItemProvider for provisional completion
-            // Calling roslyn command vscode.executeCompletionItemProvider returns null
-            // Tracked by https://github.com/dotnet/vscode-csharp/issues/7250
-            return this.provideVscodeCompletions(virtualDocument.uri, projectedPosition, triggerCharacter);
-        } finally {
-            if (virtualDocument.removeProvisionalDot()) {
-                await this.ensureProjectedCSharpDocumentUpdated(virtualDocument.uri);
-            }
-        }
+    private async delay(ms: number) {
+        return new Promise((resolve) => setTimeout(resolve, ms));
     }
 
-    private async ensureProjectedCSharpDocumentUpdated(virtualDocumentUri: vscode.Uri) {
+    private async ensureProvisionalDotUpdatedInCSharpDocument(
```

Comment:
I wonder if it would be better to hook to `vscode.workspace.onDidChangeTextDocument` to wait for the buffer to be updated. Assuming we sync all changes before artifically adding the `.`

Reply:
This makes so much sense! The reflected update is on the new commit. Thanks!